### PR TITLE
inductor(cpu): support mkldnn packed linear to improve bfloat16 performance

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96954

As title, enable mkldnn packed linear to improve bfloat16 performance.

cc @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/mkldnn.py [hunk 1]
```diff
@@ -283,52 +283,75 @@ def forward(self, input):
 
 
 class LinearUnary(nn.Linear):
-    def __init__(
-        self,
-        linear: nn.Module,
-        unary: nn.Module,
-    ):
+    def __init__(self, linear: nn.Module, unary: Optional[nn.Module], input_size: list):
```
Comment (EikanWang) [inline torch/_inductor/mkldnn.py:286]: `unary` is optional now. So does it intend to support a single `Linear`?

**Review by EikanWang — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]

File: torch/_inductor/mkldnn.py (hunk 1 above)
Comment (XiaobingSuper) [reply torch/_inductor/mkldnn.py:286]: Yes, you are right.

**Review by EikanWang — APPROVED** [review]

**Comment by XiaobingSuper** [comment]
@jansel @desertfire , please help review this PR, thanks!

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -1945,12 +1945,13 @@ def forward(self, x):
...
-            x = torch.randn(224, 224, dtype=torch.bfloat16)
-            m_opt(x)
-            self.assertEqual(m(x), m_opt(x))
+        if has_bf16_support():
+            with torch.no_grad():
+                m = M(224, 224).bfloat16().eval()
+                m_opt = torch.compile(m)
+                x = torch.randn(224, 224, dtype=torch.bfloat16)
+                m_opt(x)
+                self.assertEqual(m(x), m_opt(x))
```
Comment (desertfire) [inline test/inductor/test_torchinductor.py:1954]: Nit: is there a way to check if the weight change did happen?

**Review by desertfire — APPROVED** [review]

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (jgong5) [reply test/inductor/test_torchinductor.py:1954]: I guess we have to let users to specify the weights are constant from the frontend, something like `torch.jit.freeze`.

Comment (XiaobingSuper) [reply test/inductor/test_torchinductor.py:1954]: Yes, Let's do it at the next step.

**Pull request closed without merging**